File: problem=3/Solution.java
import java.util.Arrays;

final class Solution {

	private static void sieve(boolean[] primes) {
		Arrays.fill(primes, true);
		for (int i = 2; i < primes.length; i++) {
			if (primes[i]) {
				for (int j = (i << 1); j < primes.length; j += i) {
					primes[j] = false;
				}
			}
		}
	}

	private static int solve() {
		final long limit = 600851475143L;
		final int l = (int) Math.sqrt(limit);
		final boolean[] primes = new boolean[l + 1];

		sieve(primes);

		for (int i = primes.length - 1; i >= 0; i--) {
			if (primes[i] && limit % i == 0) {
				return i;
			}
		}

		return -1;
	}

	public static void main(String[] args) throws Exception {
		System.out.println(solve());
	}
}

File: problem=3/Solution.js
const sieve = (primes = []) => {
	primes.fill(true);
	for (let i = 2; i < primes.length; i++) {
		if (primes[i]) {
			for (let j = i << 1; j < primes.length; j += i) {
				primes[j] = false;
			}
		}
	}
};

const solve = () => {
	const limit = 600851475143;
	const l = Math.floor(Math.sqrt(limit));
	const primes = new Array(l + 1);

	sieve(primes);

	for (let i = primes.length - 1; i >= 0; i--) {
		if (primes[i] && limit % i == 0) {
			return i;
		}
	}

	return -1;
};

console.log(solve());
